apps/libsignpost/signpost_api.h

```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
 /* ENERGY API                                                             */
 /**************************************************************************/
 
+//message types
 enum energy_message_type {
     EnergyQueryMessage = 0,
     EnergyLevelWarning24hMessage = 1,
     EnergyLevelCritical24hMessage = 2,
     EnergyCurrentWarning60sMessage = 3,
+    EnergyReportMessage = 4,
+    EnergyDutyCycleMessage = 5,
 };
 
+//information sent to a module from the controller
 typedef struct __attribute__((packed)) energy_information {
-    uint32_t    energy_limit_24h_mJ;
-    uint32_t    energy_used_24h_mJ;
-    uint16_t    current_limit_60s_mA;
-    uint16_t    current_average_60s_mA;
+    uint32_t    energy_limit_mWh;
+    uint16_t    average_power_mW;
     uint8_t     energy_limit_warning_threshold;
     uint8_t     energy_limit_critical_threshold;
 } signpost_energy_information_t;
 
-_Static_assert(sizeof(signpost_energy_information_t) == 14, "On-wire structure size");
+_Static_assert(sizeof(signpost_energy_information_t) == 8, "On-wire structure size");
+
+//a mechanism for modules to report energy usage from other modules
+//this structure is for each module
+typedef struct __attribute__((packed)) energy_report_module {
+    uint8_t module_address; //module i2c address
+    uint8_t module_percent; //an integer percent 0-100 that  the module has used
```

Comment:
Okay.. to make sure I'm understanding this right, on the controller side, the situation would roughly be:
```
static radio_energy_last_time;
static radio_total_energy_used;
static modules_total_energy_used[8];

on_energy_report_module(addr, percent) {
    radio_energy_now = get_energy()
    energy_used = radio_energy_now - radio_energy_last_time;
   
   modules_total_energy_used[addr] += energy_used * percent
   radio_total_energy_used += energy_used * (1-percent)

    radio_energy_last_time = radio_energy_now
}
```

What's weird to me about this interface is that the windowing is implicit and it doesn't seem to map to operations very well. My intuition would be an interface of either:

  - Monitor operation API:
     - Starting operation on behalf of module X
     - Stopping operation on behalf of module X
     - All energy between last stop and next start charged to radio, all time between start and stop charged to X
  - Periodic API (closer to what you have):
     - At intervals send what %age of time has been spent on whom (60% self, 30% mod2, 10% mod4)

what's there now feels like some weird mix between these two approaches (you're sort of implicitly doing start/stop based on when you choose to send this message?) I have no intuition for how this assigns energy fairly or correctly to modules on the radio's behalf - I can see that it will charge some amount of energy to one module sometimes, so by that definition it's working, but I don't see how the energy that's getting charged is the correct amount

Maybe I'm just spectacularly confused